[iOS 15.3+] Crash at -[UIViewController presentViewController:withAnimationController:completion:], called on a wrong thread

eric-carlson opened the pull request:
#### 4dbd5bdfdbf72b81f0dd0f75f547df62aee50c13
<pre>
[iOS 15.3+] Crash at -[UIViewController presentViewController:withAnimationController:completion:], called on a wrong thread
<a href="https://bugs.webkit.org/show_bug.cgi?id=242235">https://bugs.webkit.org/show_bug.cgi?id=242235</a>
rdar://96305779

Reviewed by Chris Dumez.

* Source/WebCore/Modules/mediastream/MediaDevices.cpp:
(WebCore::MediaDevices::MediaDevices): Use a RunLoop::Timer as this code runs in the UI process.
* Source/WebCore/Modules/mediastream/MediaDevices.h:

* Source/WebKit/UIProcess/Cocoa/MediaPermissionUtilities.mm:
(WebKit::alertForPermission): Assert if not called on the main thread.

* Source/WebKit/UIProcess/Cocoa/UIDelegate.mm
(UIDelegate::UIClient::promptForDisplayCapturePermission): Use ensureOnMainRunLoop so we are on
the main thread when showing UI.
(UIDelegate::UIClient::decidePolicyForUserMediaPermissionRequest): Ditto.

Canonical link: <a href="https://commits.webkit.org/252189@main">https://commits.webkit.org/252189@main</a>
</pre>

cdumez commented on Source/WebKit/UIProcess/Cocoa/MediaPermissionUtilities.mm:
```diff
@@ -190,6 +190,18 @@ bool checkUsageDescriptionStringForSpeechRecognition()
 void alertForPermission(WebPageProxy& page, MediaPermissionReason reason, const WebCore::SecurityOriginData& origin, CompletionHandler<void(bool)>&& completionHandler)
 {
+    if (!isMainRunLoop()) {
+        callOnMainRunLoop([weakPage = WeakPtr { page }, reason, origin, completionHandler = WTFMove(completionHandler)]() mutable {
```
origin would need to be isolated copied.

But this is not the only issue, we just shouldn't be using a WebPageProxy from a non main thread, the `WeakPtr { page }` wouldn't be safe either.

cdumez commented on Source/WebKit/UIProcess/UserMediaPermissionRequestProxy.cpp:
```diff
@@ -151,6 +151,14 @@ void UserMediaPermissionRequestProxy::promptForGetDisplayMedia(UserMediaDisplayC
     ASSERT(m_manager && canPromptForGetDisplayMedia() && promptType != UserMediaDisplayCapturePromptType::UserChoose);
+
+    if (!isMainRunLoop()) {
+        callOnMainRunLoop([this, promptType]() {
```
You'll need to protect |this| in the lambda, otherwise, I don't see what prevents |this| from getting destroyed by the time the main run loop processes the task.

cdumez replied:
I personally would have used a ensureOnMainRunLoop() call at the call site in UIDelegate.mm. If the client is calling the completion handler off the main thread (which is a bug), then it is likely not just promptForGetDisplayMedia() that gets called off the main thread but also `allow()` and `deny()` which get called from the same completion handler.

Therefore, I suspect that this fix is only partial.

cdumez commented on Source/WebKit/UIProcess/UserMediaPermissionRequestProxy.cpp:
```diff
@@ -169,6 +177,14 @@ void UserMediaPermissionRequestProxy::promptForGetUserMedia()
     ASSERT(m_manager);
+
+    if (!isMainRunLoop()) {
+        callOnMainRunLoop([this]() {
```
Ditto.

cdumez replied:
Same comment about using ensureOnMainRunLoop() at the call site.

eric-carlson committed:
[iOS 15.3+] Crash at -[UIViewController presentViewController:withAnimationController:completion:], called on a wrong thread
https://bugs.webkit.org/show_bug.cgi?id=242235
rdar://96305779

Reviewed by Chris Dumez.

* Source/WebCore/Modules/mediastream/MediaDevices.cpp:
(WebCore::MediaDevices::MediaDevices): Use a RunLoop::Timer as this code runs in the UI process.
* Source/WebCore/Modules/mediastream/MediaDevices.h:

* Source/WebKit/UIProcess/Cocoa/MediaPermissionUtilities.mm:
(WebKit::alertForPermission): Assert if not called on the main thread.

* Source/WebKit/UIProcess/Cocoa/UIDelegate.mm
(UIDelegate::UIClient::promptForDisplayCapturePermission): Use ensureOnMainRunLoop so we are on
the main thread when showing UI.
(UIDelegate::UIClient::decidePolicyForUserMediaPermissionRequest): Ditto.

Canonical link: https://commits.webkit.org/252189@main
